### PR TITLE
[Core] Allow creating tasks/actors in a detached actor when driver has exited

### LINKED ISSUE #11437: [Multitenancy] Job ID not found in Raylet for Detached Worker
### What is the problem?
@stephanie-wang and I discovered this problem while debugging a tutorial on Serve.

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

1. `RAY_BACKEND_LOG_LEVEL=debug ray start --head`
2. Run this file as a Python script and make sure it exit
```
import ray

ray.init(address="auto")

@ray.remote
class Controller:
    def __init__(self):
        self.workers = []

    def list(self):
        return self.workers

    def start(self, actor_class):
        worker = actor_class.options(lifetime="detached").remote()
        ray.get(worker.ping.remote())
        self.workers.append(worker)

ctl = Controller.options(lifetime="detached", name="controller").remote()
print(f"Worker handles {ray.get(ctl.list.remote())}")
```
3. Run this file as a separate script, it will block forever. 
```
import ray

ray.init(address="auto")

ctl = ray.get_actor("controller")

@ray.remote
class Worker:
    def __init__(self):
        print("Worker started!")

    def ping(self): return "pong"

ray.get(ctl.start.remote(Worker))
print("worker", ray.get(ctl.list.remote()))
```
4. Now it hangs, let's ctrl+C and go to ray log and checkout raylet.out: `grep --ignore-case job raylet.out`

```
I1016 08:54:55.621362 70924 193117632 service_based_accessor.cc:92] Reestablishing subscription for job info.
I1016 08:54:55.638834 70924 193117632 service_based_accessor.cc:110] Getting all job info.
I1016 08:54:55.639565 70924 193117632 service_based_accessor.cc:117] Finished getting all job info.
I1016 08:55:01.539261 70924 193117632 worker.cc:123] Assigned worker 01000000ffffffffffffffffffffffffffffffff to job 01000000
I1016 08:55:01.539319 70924 193117632 service_based_accessor.cc:29] Adding job, job id = 01000000, driver pid = 71004
I1016 08:55:01.539764 70924 193117632 service_based_accessor.cc:39] Finished adding job, status = OK, job id = 01000000, driver pid = 71004
I1016 08:55:01.539840 70924 193117632 node_manager.cc:336] HandleJobStarted 01000000
I1016 08:55:01.546902 70924 193117632 service_based_accessor.cc:1150] Adding object location, object id = 4ea01d51bed991a789781a50e0e8dbfd00000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = e0e8dbfd
I1016 08:55:01.547286 70924 193117632 service_based_accessor.cc:1205] Removing object location, object id = 4ea01d51bed991a789781a50e0e8dbfd00000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = e0e8dbfd
I1016 08:55:01.547652 70924 193117632 service_based_accessor.cc:1166] Finished adding object location, status = OK, object id = 4ea01d51bed991a789781a50e0e8dbfd00000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = e0e8dbfd
I1016 08:55:01.548303 70924 193117632 service_based_accessor.cc:1220] Finished removing object location, status = OK, object id = 4ea01d51bed991a789781a50e0e8dbfd00000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = e0e8dbfd
I1016 08:55:01.556632 70924 193117632 service_based_accessor.cc:850] Adding task, task id = ffffffffffffffffdf5a1a8201000000, job id = 01000000
I1016 08:55:01.556875 70924 193117632 node_manager.cc:2256] Submitting task: task_spec={Type=ACTOR_CREATION_TASK, Language=PYTHON, Resources: {}, function_descriptor={type=PythonFunctionDescriptor, module_name=__main__, class_name=Controller, function_name=__init__, function_hash=8ba8d4fb-0588-4001-a7fe-2772faf99d8c}, task_id=ffffffffffffffffdf5a1a8201000000, task_name=controller:Controller.__init__, job_id=01000000, num_args=0, num_returns=1, actor_creation_task_spec={actor_id=df5a1a8201000000, max_restarts=0, max_concurrency=1, is_asyncio_actor=0, is_detached=1}}, task_execution_spec={num_forwards=0}
I1016 08:55:01.558889 70924 193117632 worker_pool.cc:325] Worker process 71007 is bound to job 01000000
I1016 08:55:01.559113 70924 193117632 service_based_accessor.cc:859] Finished adding task, status = OK, task id = ffffffffffffffffdf5a1a8201000000, job id = 01000000
I1016 08:55:01.974562 70924 193117632 worker.cc:123] Assigned worker e32a2b267032d7c9485c534d7f32e1dce8eacb4d to job 01000000
I1016 08:55:01.983006 70924 193117632 service_based_accessor.cc:1150] Adding object location, object id = eb75d8974441f778201a0155bf38150000000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = bf381500
I1016 08:55:01.983310 70924 193117632 service_based_accessor.cc:1205] Removing object location, object id = eb75d8974441f778201a0155bf38150000000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = bf381500
I1016 08:55:01.983757 70924 193117632 service_based_accessor.cc:1166] Finished adding object location, status = OK, object id = eb75d8974441f778201a0155bf38150000000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = bf381500
I1016 08:55:01.984107 70924 193117632 service_based_accessor.cc:1220] Finished removing object location, status = OK, object id = eb75d8974441f778201a0155bf38150000000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = bf381500
I1016 08:55:02.493301 70924 193117632 service_based_accessor.cc:48] Marking job state, job id = 01000000
I1016 08:55:02.493429 70924 193117632 node_manager.cc:1474] Driver (pid=71004) is disconnected. job_id: 01000000
I1016 08:55:02.493961 70924 193117632 service_based_accessor.cc:57] Finished marking job state, status = OK, job id = 01000000
I1016 08:55:02.494050 70924 193117632 node_manager.cc:353] HandleJobFinished 01000000
I1016 08:55:07.481276 70924 193117632 worker.cc:123] Assigned worker 02000000ffffffffffffffffffffffffffffffff to job 02000000
I1016 08:55:07.481319 70924 193117632 service_based_accessor.cc:29] Adding job, job id = 02000000, driver pid = 71081
I1016 08:55:07.481721 70924 193117632 service_based_accessor.cc:39] Finished adding job, status = OK, job id = 02000000, driver pid = 71081
I1016 08:55:07.481789 70924 193117632 node_manager.cc:336] HandleJobStarted 02000000
I1016 08:55:07.489056 70924 193117632 service_based_accessor.cc:1150] Adding object location, object id = 6b427e06817f752e052c96de5df7830e00000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = 5df7830e
I1016 08:55:07.489295 70924 193117632 service_based_accessor.cc:1205] Removing object location, object id = 6b427e06817f752e052c96de5df7830e00000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = 5df7830e
I1016 08:55:07.489627 70924 193117632 service_based_accessor.cc:1166] Finished adding object location, status = OK, object id = 6b427e06817f752e052c96de5df7830e00000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = 5df7830e
I1016 08:55:07.490015 70924 193117632 service_based_accessor.cc:1220] Finished removing object location, status = OK, object id = 6b427e06817f752e052c96de5df7830e00000000, node id = 16895b921f968d2a16359ac3e9de9711c51a20a1, job id = 5df7830e
I1016 08:55:07.500048 70924 193117632 service_based_accessor.cc:850] Adding task, task id = ffffffffffffffff7fffa98101000000, job id = 01000000
I1016 08:55:07.500170 70924 193117632 node_manager.cc:2256] Submitting task: task_spec={Type=ACTOR_CREATION_TASK, Language=PYTHON, Resources: {}, function_descriptor={type=PythonFunctionDescriptor, module_name=__main__, class_name=Worker, function_name=__init__, function_hash=c66af86d-070e-4c79-a178-29e8ac074eb6}, task_id=ffffffffffffffff7fffa98101000000, task_name=Worker.__init__(), job_id=01000000, num_args=0, num_returns=1, actor_creation_task_spec={actor_id=7fffa98101000000, max_restarts=0, max_concurrency=1, is_asyncio_actor=0, is_detached=1}}, task_execution_spec={num_forwards=0}
I1016 08:55:07.500252 70924 193117632 worker_pool.cc:180] Job config of job 01000000 are not local yet.
I1016 08:55:07.500522 70924 193117632 service_based_accessor.cc:859] Finished adding task, status = OK, task id = ffffffffffffffff7fffa98101000000, job id = 01000000
I1016 08:55:15.822947 70924 193117632 service_based_accessor.cc:1378] Publishing job error, job id = 01000000
I1016 08:55:15.822988 70924 193117632 service_based_accessor.cc:1381] Finished publishing job error, job id = 01000000
I1016 08:59:19.770591 70924 193117632 service_based_accessor.cc:48] Marking job state, job id = 02000000
I1016 08:59:19.770685 70924 193117632 node_manager.cc:1474] Driver (pid=71081) is disconnected. job_id: 02000000
I1016 08:59:19.771344 70924 193117632 service_based_accessor.cc:57] Finished marking job state, status = OK, job id = 02000000
I1016 08:59:19.771391 70924 193117632 node_manager.cc:353] HandleJobFinished 02000000
```

Notice the 
```
I1016 08:55:07.500252 70924 193117632 worker_pool.cc:180] Job config of job 01000000 are not local yet.
```

If we cannot run your script, we cannot fix your issue.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

Previously, when a driver exits, the job config in `WorkerPool::unfinished_jobs_` will be cleared. This is to avoid registering leaked workers after the driver exits (https://github.com/ray-project/ray/pull/10569). But as described in #11437, we should allow creating new tasks/actors from a detached actor even if the driver of the detached actor has exited.

This PR changes `unfinished_jobs_` to `all_jobs_`, and keep the job config in `all_jobs_` when the driver exits.

## Related issue number

Closes #11437

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kfstorm** [comment]
@rkooo567 I'm not sure why, but `test_worker_registration_failure_after_driver_exit` can pass on my local environment.

**Comment by rkooo567** [comment]
@kfstorm Can you explain the drawback of this approach again? If I understood correctly, this means tasks/actors can be scheduled even after the job is terminated. What're the worst consequences of that behavior?

**Comment by kfstorm** [comment]
@rkooo567 No, actors and tasks won't be scheduled after the job is terminated because here we kill all workers without detached workers of this job.

https://github.com/ray-project/ray/blob/2fc323704557191eacbef435f78d7014a74d8587/src/ray/raylet/node_manager.cc#L352-L370

However, if there are tasks/actors that are being scheduled, the worker pool may spawn new worker processes for these actors. If the job is terminated after worker processes are spawned and before they are registered, these workers will remain alive because they aren't running tasks/actors yet.

If there are other jobs running in the cluster, we hope that the worker capping feature will clean up these idle workers later.

**Comment by rkooo567** [comment]
@kfstorm I am a little confused what's the difference between these two sentences;

> actors and tasks won't be scheduled after the job is terminated

> If there are tasks/actors that are being scheduled,

Does the second sentence mean tasks/actors that are queued to be scheduled?

File: src/ray/raylet/worker_pool.cc [hunk 1]
```diff
@@ -427,11 +427,12 @@ void WorkerPool::MarkPortAsFree(int port) {
...
 
 void WorkerPool::HandleJobStarted(const JobID &job_id, const rpc::JobConfig &job_config) {
-  unfinished_jobs_[job_id] = job_config;
+  all_jobs_[job_id] = job_config;
 }
 
 void WorkerPool::HandleJobFinished(const JobID &job_id) {
-  unfinished_jobs_.erase(job_id);
+  // Currently we don't erase the job from `all_jobs_` , as a workaround for
+  // https://github.com/ray-project/ray/issues/11437.
```
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:435]: Can we just keep `unfinished_jobs_.erase(job_id);` commented in?

**Comment by kfstorm** [comment]
@rkooo567 Maybe the below timeline can help you understand.

1. The driver submits a task.
2. Raylet receives the task and spawned a worker process.
3. The driver exits.
4. Raylet noticed that the driver has exited, kills all registered workers of this job, and removes all tasks of this job from the local task queue.
5. The newly spawned worker process tries to register to Raylet, and Raylet accepts.
6. Since there isn't any task of this job to be scheduled, the newly spawned worker will remain idle.
7. The newly spawned worker process stays alive until worker capping tries to terminate it.

I created https://github.com/ray-project/ray/pull/10569 to solve this issue by rejecting further worker registration once the job is terminated. But now as a temporary fix to allow detached actors to create tasks/actors, this PR has to allow further worker registrations even if the job is terminated.

**Comment by rkooo567** [comment]
Gotcha. It makes sense. Thanks for the clarification :). So the only possible race condition is (worker process spawned -> driver exits -> register reply comes back). 

That being said, `test_worker_registration_failure_after_driver_exit ` sounds indeed weird. When I looked at tests, it looks like it has to be deleted. Do you know what's the value of `get_workers()` in `wait_for_condition(lambda: len(get_workers()) == before)` after this PR? (maybe there is another code block that keeps workers from being killed by worker capping mechanism)?

**Comment by rkooo567** [comment]
Btw, how's this scenario covered? 

1. Driver exits and tasks being removed from the queue.
2. But there's a task that is received from the driver that isn't queued yet.
3. After 1, this task is queued.

**Comment by kfstorm** [comment]
Ah, I see. `test_worker_registration_failure_after_driver_exit` can pass because we have worker capping enabled. If I disable worker capping, the test will fail.

**Comment by rkooo567** [comment]
That makes sense! Then it looks like a great solution!

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
There's one concern here. Windows test fails with `serve:test_api`. Let me re-run this test.

**Comment by kfstorm** [comment]
> Btw, how's this scenario covered? (Is it even possible?)
> 
> 1. Driver exits and tasks being removed from the queue.
> 2. But there's a task that is received from the driver that isn't queued yet.
> 3. After 1, this task is queued.

Great question. I don't think we have any kind of check for this currently. With this PR, we do allow creating tasks from a detached actor, so that must mean the scenario you mentioned is allowed in Ray.

**Comment by rkooo567** [comment]
Maybe we can track this as a possible issue? I think in this scenario, this can happen. 

1. Driver in node A sends a task schedule request R to node B.
2. Driver right after instantly.
3. node B receives GCS notification before R is sent to node B (due to some network delay or something).

**Comment by kfstorm** [comment]
@rkooo567 It makes sense to me.

File: src/ray/raylet/worker_pool.cc (hunk 1 above)
Comment (kfstorm) [reply src/ray/raylet/worker_pool.cc:435]: done

Comment (stephanie-wang) [reply src/ray/raylet/worker_pool.cc:435]: Maybe change to `// all_jobs_.erase(job_id);`?

**Review by stephanie-wang — APPROVED** [review]
Thanks, @kfstorm!

**Pull request merged**